File: java/challenges/lib/src/main/java/challenges/stacksandqueues/Queue.java
package challenges.stacksandqueues;

import java.util.EmptyStackException;
import java.util.NoSuchElementException;

public class Queue {
  public Node front;
  public Node rear;

  public Queue(){
    this.front = null;
    this.rear = null;
  }

  public void enqueue(Object val) {
    Node newNode = new Node(val);
    if (isEmpty()) {
      front = newNode;
    } else {
      rear.next = newNode;
    }
    rear = newNode;
  }

  public Object dequeue(){
    if(front == null){
      throw new NoSuchElementException();
    }
    Node temp = front;
    front = front.next;
    temp.next = null;

    return temp.val;
  }

  public Object peek(){
    if(front == null){
      throw new NoSuchElementException();
    }
    return front.val;
  }

  public boolean isEmpty(){
    return front == null;
  }

}

File: javascript/stackAndQueue/queue.js
'use strict';

class Node {
  constructor(val)
  {
    this.val = val;
    this.next = null;
  }
}

class Queue {
  constructor()
  {
    this.front = null;
    this.back = null;
  }

  enqueue(val) {
    let newNode = new Node(val);

    if (this.isEmpty()) {
      this.front = newNode;
    } else {
      this.front.next = newNode;
    }
    this.back = newNode;
  }

  dequeue() {
    if (!this.front) {
      return null;
    }
    let temp = this.front;
    this.front = this.front.next;
    temp.next = null;

    return temp.val;
  }

  peek() {
    return this.front.val;
  }

  isEmpty() {
    return this.front === null;
  }

}

module.exports = { Queue, Node };
